### ISSUE TITLE
Wrong result about visualization

### ISSUE THREAD

**Issue by SWRDFK** [wangguanan/Pytorch-Person-REID-Baseline-Bag-of-Tricks#2]
Hi, author, thank you for your excellent work. I add some visualization code in the file "retrieval.py" to show the top 10 similar images with the query image by the method "evaluate". But I got the wrong result as the following picture. It seems something wrong with the pids of gallery images which are different from each other. To say, the labels are the same but not the correct ones, so I think the labeling process or the "evaluate" method are not right. 

![图片](https://user-images.githubusercontent.com/40492484/58928514-5d772400-8785-11e9-830a-91368b0e20a3.png)

**Comment by wangguanan** [comment]
Hi, thanks for your advice. I will check it as soon as possible. Could you provide more information?

**Comment by SWRDFK** [comment]
Thank you so much for your reply! I implement the visualization like this code:  [https://github.com/layumi/Person_reID_baseline_pytorch/blob/master/demo.py](url)

In order to get one query image, I use their form of datasets “/Market-1501-v15.09.15/pytorch/”
“image_datasets = {x: datasets.ImageFolder(f'/home/xxxx/Market-1501-v15.09.15/pytorch/{x}/') for x in ['gallery','query']}”

Besides, I modify the method evaluate() as following:
`   def evaluate(self, query_feature, query_cam, query_label, gallery_feature, gallery_cam, gallery_label):
        '''
        :param query_feature: np.array, 1d
        :param query_cam: int
        :param query_label: int
        :param gallery_feature: np.array, 2d, gallerys_size * feature_dim
        :param gallery_cam: np.array, 1d
        :param gallery_label: np.array, 1d
        :return:
        '''

        # cosine score
        if self.dist is 'cosine':
            # feature has been normalize during intialization
            score = np.matmul(query_feature, gallery_feature.transpose())
            index = np.argsort(score)[::-1]
        elif self.dist is 'euclidean':
            score = self.l2(query_feature.reshape([1, -1]), gallery_feature)
            index = np.argsort(score.reshape([-1]))

        junk_index_1 = self.in1d(np.argwhere(query_label == gallery_label), np.argwhere(query_cam == gallery_cam))
        junk_index_2 = np.argwhere(gallery_label == -1)
        junk_index = np.append(junk_index_1, junk_index_2)

        good_index = self.in1d(np.argwhere(query_label == gallery_label), np.argwhere(query_cam != gallery_cam))
        index_wo_junk = self.notin1d(index, junk_index)

        index = index[index_wo_junk]

        # return self.compute_AP(index_wo_junk, good_index)
        return index_wo_junk, good_index, index`

As a result, I got wrong images to the given 2000th query image. The number behind the root is label of image, which is different from its true label.
![图片](https://user-images.githubusercontent.com/40492484/59548548-941e1d00-8f83-11e9-974e-ed9b67f6b9b8.png)
![图片](https://user-images.githubusercontent.com/40492484/59548589-0ee73800-8f84-11e9-95a1-5c57a627482c.png)

And this is the result of their demo:
![图片](https://user-images.githubusercontent.com/40492484/59548617-553c9700-8f84-11e9-9f6d-6b38d11d2b89.png)
![图片](https://user-images.githubusercontent.com/40492484/59548580-f9720e00-8f83-11e9-8ca0-c2808d4caad8.png)

Thank you again for your patience, please contact me if you want to know more details about this problem.

**Comment by wangguanan** [comment]
Hi, you mean that although the quantitative scores (rank1, map) are very high, the visualization of the rank list is totally wrong, when you use the visualization code and data_loader code from layumi, and the training code from this project?

The data_loader codes of ours and layumi are different. if you use layumi's data_loader code to get query images, you should also use his code to get gallery images. could you please check it first?

**Comment by SWRDFK** [comment]
OK, I got it but not exactly know how to get images by your data_loader, should I implement this in core/test.py?

**Comment by wangguanan** [comment]
1. Our data_loader code is in core/data_loader.
2. I guess that maybe the shuffled test_loader causes to this problem, you can try by setting shuffle=False.
3. If 2 doesn't work, you can try our new code, where we have added our visualization code to it. It seems that nothing wrong with our visualization code.
4. acknowledge: our visualization code is based on deep-person-reid library (https://github.com/KaiyangZhou/deep-person-reid)

**Comment by SWRDFK** [comment]
Thank you, author. Your visualization code works well. I have learned a lot from you and I'll work harder.

**Comment by wangguanan** [comment]
You are welcome.
